sys/kern/src/task.rs

```diff
@@ -221,12 +221,44 @@ impl Task {
             // according to the task's region map... but fine with us.
             return true;
         }
+        // Forceably include DEVICE in the forbidden set, whether or not the
+        // caller thought about it.
         let forbidden = forbidden | RegionAttributes::DEVICE;
-        self.region_table().iter().any(|region| {
-            region.covers(slice)
-                && region.attributes.contains(desired)
-                && !region.attributes.intersects(forbidden)
-        })
+
+        // We need to be convinced that this slice is _entirely covered_ by
+        // regions with the right attributes. It _may_ span multiple regions,
+        // which is important since the build system can insert region
+        // boundaries in unpredictable locations in otherwise innocent programs.
+        // But the regions that are spanned must be perfectly contiguous.
+
+        let mut scan_addr = slice.base_addr;
+        let end_addr = slice.end_addr();
+
+        for region in self.region_table() {
```

Comment:
A comment here explaining that `region_table` is (1) sorted by start address and (2) non-overlapping would help clarify this loop

Reply:
Adding!